features/nanostack/FEATURE_NANOSTACK/nanostack-interface/NanostackInterface.cpp

```diff
@@ -524,104 +516,144 @@ nsapi_error_t NanostackInterface::socket_close(void *handle)
-                                       (uint8_t *)data, size);
-        /*
-         * \return 0 on success.
-         * \return -1 invalid socket id.
-         * \return -2 Socket memory allocation fail.
-         * \return -3 TCP state not established.
-         * \return -4 Socket tx process busy.
-         * \return -5 TLS authentication not ready.
-         * \return -6 Packet too short.
-         * */
-        if (-4 == send_to_status) {
-            ret = NSAPI_ERROR_WOULD_BLOCK;
-        } else if (0 != send_to_status) {
-            tr_error("socket_sendto: error=%d", send_to_status);
-            ret = NSAPI_ERROR_DEVICE_ERROR;
-        } else {
-            ret = size;
-        }
-    }
+    int retcode;
+#if 0
+    retcode = ::socket_sendto(socket->socket_id, address,
+                                            data, size);
+#else
+    // Use sendmsg purely to get the new return style
+    // of returning data written rather than 0 on success,
+    // which means TCP can do partial writes. (Sadly,
+    // it's the only call which takes flags so we can
+    // leave the NS_MSG_LEGACY0 flag clear).
+    ns_msghdr_t msg;
+    ns_iovec_t iov;
+    iov.iov_base = const_cast<void *>(data);
+    iov.iov_len = size;
+    msg.msg_name = const_cast<ns_address_t *>(address);
+    msg.msg_namelen = address ? sizeof *address : 0;
+    msg.msg_iov = &iov;
+    msg.msg_iovlen = 1;
+    msg.msg_control = NULL;
+    msg.msg_controllen = 0;
+    retcode = ::socket_sendmsg(socket->socket_id, &msg, 0);
+#endif
 
     nanostack_unlock();
 
+    /*
+     * \return length if entire amount written (which could be 0)
+     * \return value >0 and <length if partial amount written (stream only)
+     * \return NS_EWOULDBLOCK if nothing written due to lack of queue space.
+     * \return -1 Invalid socket ID or message structure.
+     * \return -2 Socket memory allocation fail.
+     * \return -3 TCP state not established or address scope not defined .
+     * \return -4 Socket TX process busy or unknown interface.
+     * \return -5 Socket not connected
+     * \return -6 Packet too short (ICMP raw socket error).
+     * */
+    if (retcode == -2 || retcode == NS_EWOULDBLOCK) {
+        ret = NSAPI_ERROR_WOULD_BLOCK;
+    } else if (retcode < 0) {
+        tr_error("socket_sendmsg: error=%d", retcode);
+        ret = NSAPI_ERROR_DEVICE_ERROR;
+    } else {
+        ret = retcode;
+    }
+
+out:
     tr_debug("socket_sendto(socket=%p) sock_id=%d, ret=%i", socket, socket->socket_id, ret);
 
     return ret;
 }
 
+nsapi_size_or_error_t NanostackInterface::socket_sendto(void *handle, const SocketAddress &address, const void *data, nsapi_size_t size)
+{
+    if (address.get_ip_version() != NSAPI_IPv6) {
+        return NSAPI_ERROR_UNSUPPORTED;
+    }
+
+    ns_address_t ns_address;
+    convert_mbed_addr_to_ns(&ns_address, &address);
+
+    return do_sendto(handle, &ns_address, data, size);
```

Comment:
might add a comment that no locking is needed in socket_sendto since do_sendto is thread safe.